### PR TITLE
Country key download and exposure detection  (Interop – 2667, 2764)

### PR DESCRIPTION
**Pull request by ArturFriesen**
## Description

This PR introduces the download of key packages based on the country approach. All country key packages are combined for exposure detection.

Story: [2764](https://jira.itc.sap.com/browse/EXPOSUREAPP-2764)

## Checklist

* [x] This PR does not change text that hasn't been signed off with the RKI. Have a look at the pinned issue [440](https://github.com/corona-warn-app/cwa-app-ios/issues)
* [x] Test your changes as thoroughly as possible before you commit them. Preferably, automate your test by unit/integration tests.
* [ ] If this PR comes from a fork, please [Allow edits from maintainers](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork)
* [x] Set a speaking title. Format: {task_name} (closes #{issue_number}). For example: Use logger (closes # 41)
* [x] [Link your Pull Request to an issue](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) - Pull Requests that are not linked to an issue with you as an assignee will be closed.
* [ ] Create Work In Progress [WIP] pull requests only if you need clarification or an explicit review before you can continue your work item.
* [x] Make sure that your PR is not introducing _unnecessary_ reformatting (e.g., introduced by on-save hooks in your IDE)

### CODE REVIEW

File: src/xcode/ENA/ENA/Source/Client/HTTP Client/HTTPClient.swift [hunk 1]
```diff
@@ -213,6 +196,23 @@ final class HTTPClient: Client {
...
+		appConfiguration { config in
+			guard let config = config else {
+				completion(nil)
+				return
+			}
+			guard config.hasExposureConfig else {
+				completion(nil)
+				return
+			}
+			completion(try? ENExposureConfiguration(from: config.exposureConfig, minRiskScore: config.minRiskScore))
```
Comment (johannesrohwer) [inline src/xcode/ENA/ENA/Source/Client/HTTP Client/HTTPClient.swift:212]: the docs in the protocol say the completion handler will always be called on the main thread. is this guaranteed here?

File: src/xcode/ENA/ENA/Source/Client/HTTP Client/__tests__/HTTPClient+AppConfigTests.swift [hunk 2]
```diff
@@ -97,6 +97,9 @@ final class HTTPClientAppConfigTests: XCTestCase {
 			case .failure(let error):
 				XCTFail("Country fetch error: \(error as NSError)")
 			case .success(let list):
+				
+				// As of 2020-09-16 the latest backend deployment returns zero (0) countries.
+				// Due to the current changes in INTEROP we should check this by end of Sept 2020!
```
Comment (johannesrohwer) [inline src/xcode/ENA/ENA/Source/Client/HTTP Client/__tests__/HTTPClient+AppConfigTests.swift:102]: should we add a ticket so we do not forget about this?

File: src/xcode/ENA/ENA/Source/Services/DownloadedPackagesStore/V1/DownloadedPackagesStoreV1.swift [hunk 3]
```diff
@@ -21,12 +21,12 @@ import FMDB
...
-	func package(for day: String, country: String) -> SAPDownloadedPackage?
-	func hourlyPackages(for day: String, country: String) -> [SAPDownloadedPackage]
-	func allDays(country: String) -> [String] // 2020-05-30
-	func hours(for day: String, country: String) -> [Int]
+	func set(country: Country.ID, day: String, package: SAPDownloadedPackage)
+	func set(country: Country.ID, hour: Int, day: String, package: SAPDownloadedPackage)
+	func package(for day: String, country: Country.ID) -> SAPDownloadedPackage?
+	func hourlyPackages(for day: String, country: Country.ID) -> [SAPDownloadedPackage]
+	func allDays(country: Country.ID) -> [String] // 2020-05-30
+	func hours(for day: String, country: Country.ID) -> [Int]
```
Comment (johannesrohwer) [inline src/xcode/ENA/ENA/Source/Services/DownloadedPackagesStore/V1/DownloadedPackagesStoreV1.swift:29]: Country.ID >>> String 🎉

File: src/xcode/ENA/ENA/Source/Services/Exposure Transaction/CountryKeypackageDownloader.swift [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+			return
+		}
+
+		guard let deltaDaysAndHours = delegate?.exposureDetection(country: country, downloadDeltaFor: daysAndHours) else {
+			completion(.failure(.noDaysAndHours))
+			return
+		}
+
+		delegate?.exposureDetection(country: country, downloadAndStore: deltaDaysAndHours) { error in
+			if error != nil {
```
Comment (johannesrohwer) [inline src/xcode/ENA/ENA/Source/Services/Exposure Transaction/CountryKeypackageDownloader.swift:57]: nitpick: guard

File: src/xcode/ENA/ENA/Source/Services/Exposure Transaction/ExposureDetection.swift [hunk 5]
```diff
@@ -25,53 +25,104 @@ final class ExposureDetection {
...
-		guard let writtenPackages = delegate?.exposureDetectionWriteDownloadedPackages(self) else {
-			endPrematurely(reason: .unableToWriteDiagnosisKeys)
-			return
+
+		dispatchGroup.notify(queue: .main) {
+			if let error = errors.first {
+				self.endPrematurely(reason: error)
+			} else {
+				completion()
+			}
```
Comment (johannesrohwer) [inline src/xcode/ENA/ENA/Source/Services/Exposure Transaction/ExposureDetection.swift:91]: why do we collect all errors and only return the first one?

File: src/xcode/ENA/ENA/Source/Services/Exposure Transaction/__tests__/ExposureDetectionTests.swift [hunk 6]
```diff
@@ -75,18 +89,67 @@ final class ExposureDetectionTransactionTests: XCTestCase {
...
+		let expectationDetectionCompletion = expectation(description: "Detection completion was called.")
+		detection.start { _ in
+			expectationDetectionCompletion.fulfill()
+		}
+
+		waitForExpectations(timeout: 1.0)
+	}
+
+	func makeCountries() -> [Country] {
+		guard let enCountry = Country(countryCode: "FR"),
```
Comment (johannesrohwer) [inline src/xcode/ENA/ENA/Source/Services/Exposure Transaction/__tests__/ExposureDetectionTests.swift:145]: nitpick: frCountry? 
could also be shortened to something like:
`[Country(countryCode: "FR"), Country(countryCode: "IT")].compactMap { $0 }`
(which does not XCFail() when the country init fails but this should be covered by other tests..)

File: src/xcode/ENA/ENA/Source/Services/Risk/Provider/RiskProvider.swift [hunk 7]
```diff
@@ -254,7 +254,7 @@ extension RiskProvider: RiskProviding {
 			group.leave()
 		}
 
-		guard group.wait(timeout: .now() + .seconds(60)) == .success else {
+		guard group.wait(timeout: .now() + .seconds(60 * 8)) == .success else {
```
Comment (johannesrohwer) [inline src/xcode/ENA/ENA/Source/Services/Risk/Provider/RiskProvider.swift:257]: is any reasoning why 8 min is the perfect amount of time to wait?

File: src/xcode/ENA/ENA/Source/Services/Exposure Transaction/ExposureDetection.swift [hunk 8]
```diff
@@ -25,53 +25,104 @@ final class ExposureDetection {
...
 			}
-			self.delegate?.exposureDetection(self, downloadConfiguration: self.useConfiguration)
-		}
+		})
+	}
+
+	private func getCountriesToDetect(supportedCountries: [Country]) -> [Country.ID] {
+		var countryIDs = supportedCountries.map { $0.id }
+		countryIDs.append(Country.defaultCountry().id)
+		return countryIDs
```
Comment (johannesrohwer) [inline src/xcode/ENA/ENA/Source/Services/Exposure Transaction/ExposureDetection.swift:64]: would it make sense to eliminate possible duplicates here? its unlikely but if we get a dup here (e.g. the default country) we might run into unexpected trouble somewhere else.

**Review by johannesrohwer — CHANGES_REQUESTED** [review]
awesome PR 🥇 just a few minor points from my side!

File: src/xcode/ENA/ENA/Source/Client/HTTP Client/__tests__/HTTPClient+AppConfigTests.swift (hunk 2 above)
Comment (ArturFriesen) [reply src/xcode/ENA/ENA/Source/Client/HTTP Client/__tests__/HTTPClient+AppConfigTests.swift:102]: Added subtask.https://jira.itc.sap.com/browse/EXPOSUREAPP-2783

File: src/xcode/ENA/ENA/Source/Services/DownloadedPackagesStore/V1/DownloadedPackagesStoreV1.swift (hunk 3 above)
Comment (ArturFriesen) [reply src/xcode/ENA/ENA/Source/Services/DownloadedPackagesStore/V1/DownloadedPackagesStoreV1.swift:29]: 🥳

File: src/xcode/ENA/ENA/Source/Services/Exposure Transaction/CountryKeypackageDownloader.swift (hunk 4 above)
Comment (ArturFriesen) [reply src/xcode/ENA/ENA/Source/Services/Exposure Transaction/CountryKeypackageDownloader.swift:57]: added guard

File: src/xcode/ENA/ENA/Source/Services/Exposure Transaction/ExposureDetection.swift (hunk 8 above)
Comment (ArturFriesen) [reply src/xcode/ENA/ENA/Source/Services/Exposure Transaction/ExposureDetection.swift:64]: Good idea. Used a Set to do this.

File: src/xcode/ENA/ENA/Source/Services/Exposure Transaction/ExposureDetection.swift (hunk 5 above)
Comment (ArturFriesen) [reply src/xcode/ENA/ENA/Source/Services/Exposure Transaction/ExposureDetection.swift:91]: Its a dispatch group and the UI can only handle one error. If one error occurs, risk calculation will not be triggered.

Comment (ArturFriesen) [reply src/xcode/ENA/ENA/Source/Services/Exposure Transaction/ExposureDetection.swift:91]: I have added 2 tests for this unhappy path.

File: src/xcode/ENA/ENA/Source/Services/Exposure Transaction/__tests__/ExposureDetectionTests.swift (hunk 6 above)
Comment (ArturFriesen) [reply src/xcode/ENA/ENA/Source/Services/Exposure Transaction/__tests__/ExposureDetectionTests.swift:225]: changed

File: src/xcode/ENA/ENA/Source/Services/Risk/Provider/RiskProvider.swift (hunk 7 above)
Comment (ArturFriesen) [reply src/xcode/ENA/ENA/Source/Services/Risk/Provider/RiskProvider.swift:257]: Its a requirement from the story.

File: src/xcode/ENA/ENA/Source/Client/HTTP Client/HTTPClient.swift (hunk 1 above)
Comment (ArturFriesen) [reply src/xcode/ENA/ENA/Source/Client/HTTP Client/HTTPClient.swift:212]: This comment is wrong. I removed it.
This function is only used by the ExposureDetection.
The ExposureDetection "forwards" this call to the UI on the main thread.

**Review by melloskitten — APPROVED** [review]
Thanks for your contribution. Looks good to me :-)

**Review by johannesrohwer — APPROVED** [review]
Thank you for implementing the feedback! Awesome contribution 🎉

**Pull request merged**